### PR TITLE
SE-0067 - Failable initializers for Numeric Types

### PR DESCRIPTION
**Pull request by ultramiraculous**
#### What's in this pull request?

Adding failable initializers for builtin numeric types. Added `init?` methods to each numeric type and moved the Float->Int conversions into the FixedPoint template.
#### Resolved bug number: ([SR-1491](https://bugs.swift.org/browse/SR-1491))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by ultramiraculous** [comment]
Is there a test suite around this stuff that I'm missing?

File: stdlib/public/core/FloatingPoint.swift.gyb [hunk 1]
```diff
@@ -75,18 +75,9 @@ public protocol FloatingPoint: Comparable, IntegerLiteralConvertible,
...
-  /*  TODO: Implement the following APIs once a revised integer protocol is
-      introduced that allows for them to be implemented.  In particular, we
-      need to have an "index of most significant bit" operation and "get
-      absolute value as unsigned type" operation on the Integer protocol.
-
-  /// The closest representable value to the argument.
-  init<Source: Integer>(_ value: Source)
-
-  /// Fails if the argument cannot be exactly represented.
-  init?<Source: Integer>(exactly value: Source)
```
Comment (gribozavr) [inline stdlib/public/core/FloatingPoint.swift.gyb]: Please restore this comment.  The initializers should be generic.

**Comment by gribozavr** [comment]
Thank you!  The tests should go into `validation-test/stdlib/FixedPoint.swift.gyb`, `validation-test/stdlib/FixedPointArithmeticTraps.swift.gyb`, `test/1_stdlib/FloatingPoint.swift.gyb`.  Feel free to create a new file in the validation test suite for floating point if necessary.

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@ultramiraculous Could you take a look at the build failures?  Do you think they are related?

**Comment by ultramiraculous** [comment]
They don't _appear_ to be related to my changes, but I've said that before and learned my lesson. It looks like there's unused results in Foundation? (Was SE-0047 completed at some point?)

**Comment by gribozavr** [comment]
Quite possibly so, but unused results should only be warnings.  In this case, we're failing the swiftpm bootstrap on Linux.

**Comment by ultramiraculous** [comment]
Just got around to taking another look. It looks like the OS X test failed for a different reason than the Linux one:

```
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/stdlib/public/core/FloatingPointTypes.swift.gyb:740:18: error: integer literal '4503599627370495' overflows when stored into 'UInt'
    if (value) > 4503599627370495 && UInt(self) != value {
```

Which is clearly my fault. I'll look at a fix.

**Comment by ultramiraculous** [comment]
I switched `FloatingPointTypes` to use `all_integer_types` like `FixedPoint`. I didn't notice 'Int' was in `FloatingPointTypes`'s `allIntBits`, and "Int" > $a_number is always true, which screwed up my logic.

I have tests mostly done, but is there a good way to build + test just the stdlib? I'm relying on hacked up build scripts to try to validate my work in a reasonable timeframe, which I imagine isn't the right way to do this?

**Comment by gribozavr** [comment]
Thank you!  The usual way to test is to run `build-script -RT`.

**Comment by gribozavr** [comment]
We can also run the tests on CI.

@swift-ci Please test

**Comment by ultramiraculous** [comment]
Added some tests to try to hit some edge cases.

**Comment by ultramiraculous** [comment]
Still not sure what to make of the build failures, any insight @gribozavr?

**Comment by moiseev** [comment]
Since the time it was last looked at, there are some merge conflicts. @ultramiraculous could you please resolve them?

**Comment by moiseev** [comment]
@swift-ci Please test

File: validation-test/stdlib/FixedPointConversion.swift.gyb [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+//
+// RUN: %S/../../utils/line-directive %t/FixedPointConversion.swift -- %target-run %t/a.out_Debug
+// RUN: %S/../../utils/line-directive %t/FixedPointConversion.swift -- %target-run %t/a.out_Release
+// REQUIRES: executable_test
+
+%{
+
+from SwiftIntTypes import all_integer_types
+
+word_bits = 4
```
Comment (moiseev) [inline validation-test/stdlib/FixedPointConversion.swift.gyb]: This doesn't look right.

File: stdlib/public/core/FixedPoint.swift.gyb [hunk 3]
```diff
@@ -490,6 +505,40 @@ extension ${Self} {
...
+    _precondition(value > ${str(lower)}.0,
+      "${Src} value cannot be converted to ${Self} because the result would be less than ${Self}.min")
+    _precondition(value < ${str(upper)}.0,
+      "${Src} value cannot be converted to ${Self} because the result would be greater than ${Self}.max")
+    self._value = Builtin.fpto${sign}i_FPIEEE${srcBits}_${BuiltinName}(value._value)
+  }
+
+  /// Creates a ${Self} whose value is `value` rounded towards zero.
+  @_transparent
+  public init?(exactly value: ${Src}) {
```
Comment (moiseev) [inline stdlib/public/core/FixedPoint.swift.gyb]: `Src` here only resolves to floating point types. But in the tests later it is being called as if it were `Int.init?(exactly: Int)`.

File: validation-test/stdlib/FixedPointConversion.swift.gyb [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  let result = ${Self}(input)
+  expectEqual(${testValue}, result)
+  _blackHole(result)
+}
+
+/// Never-nil failable conversion from ${Other}(${testValue}) to ${Self}.
+FixedPointConversionFailure.test("${Other}To${Self}FailableConversion/dest=${testValue}") {
+  // Test that nothing interesting happens and we end up with a non-nil, identical result.
+  let input = get${Other}(${testValue})
+  var result = ${Self}(exactly: input)
```
Comment (moiseev) [inline validation-test/stdlib/FixedPointConversion.swift.gyb]: Here both `Self` and `Other` are integer types, and there are no initializers with `exact` argument label on integers that accept other integers.

File: validation-test/stdlib/FixedPointConversion.swift.gyb [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+%       end
+
+%     end # for in testValues
+
+%     if Other == 'Float80':
+#endif
+%     end
+
+%   end # for in floatNameToSignificandBits (Other)
+% end # for in all_integer_types (Self)
```
Comment (moiseev) [inline validation-test/stdlib/FixedPointConversion.swift.gyb]: This test file misses `runAllTests()` call at the very end.

File: stdlib/public/core/FixedPoint.swift.gyb (hunk 3 above)
Comment (ultramiraculous) [inline stdlib/public/core/FixedPoint.swift.gyb]: I actually just noticed this looking at the build failures. I somehow lost a commit when rebasing.

Comment (moiseev) [inline stdlib/public/core/FixedPoint.swift.gyb]: :+1:

Comment (moiseev) [inline stdlib/public/core/FixedPoint.swift.gyb]: Please let me know when you include the missing commit.

Comment (ultramiraculous) [inline stdlib/public/core/FixedPoint.swift.gyb]: Found it + added it back. Int->Int failable initializers are back.

Also added the runAllTests()

**Comment by ultramiraculous** [comment]
@moiseev / @gribozavr Would it make sense to just break this up into 2 or 4 PRs? It got a bit larger than I'd usually prefer to have reviewed after adding the tests, and maybe we could at least get the stable parts of it merged in.

**Comment by gribozavr** [comment]
@ultramiraculous Absolutely, splitting the PR would make things easier!  I just have one small request -- while splitting, could you fold the tests for each incremental step into PRs that add or change the APIs?

**Pull request closed without merging**